### ISSUE TITLE
docker images are only available for amd64

### ISSUE THREAD

**Issue by allexoll** [cross-rs/cross#645]
Currently all docker image released for cross are amd64. This means that all other host platform will have to run it trough an emulation layer, which slows it down by a lot, where if the images were published for multiple arch, docker would choose it based on the host arch.

example: 

Hos: *aarch64-apple-darwin*
```
> uname -a
Darwin laptop.local 21.3.0 Darwin Kernel Version 21.3.0: Wed Jan  5 21:37:58 PST 2022; root:xnu-8019.80.24~20/RELEASE_ARM64_T6000 arm64
```

Target I want to compile for: *aarch64-unknown-linux-gnu*
```
> docker run --rm -it rustembedded/cross:aarch64-unknown-linux-gnu "sh" -c  "uname -a"
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
Linux 23858603f374 5.10.76-linuxkit #1 SMP PREEMPT Mon Nov 8 11:22:26 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

which is not very efficient when it works, if at all, because the whole thing is run trough qemu_x86_64 for arm.

I'm not sure how easy it would be to build & publish the same set of images for other arch, but i think that's an issue that other docker images publisher have solved?

I'd be happy to look into if if that support is a wanted enhancement when i have the time, but i'd happily take pointers on how to test the potential solution.

Best,

allexoll

**Comment by Emilgardis** [comment]
Thanks for the issue! I think we can solve this quite easily with docker buildx.

**Comment by allexoll** [comment]
I don't think it is that easy... Some dockerfiles also install libc6-dev cross and g++ cross, which would be inexistant if the mapping is host arch == target arch.

For example if i want to build a aarch64 image for a aarch64 host, 
```
RUN apt-get install --assume-yes --no-install-recommends \
    g++-aarch64-linux-gnu \
    libc6-dev-arm64-cross
```

That would fail because they are not available. QEMU would also not be needed in this case, and the linux-runner also intends to be running on x86_64 or i?86, otherwise it runs trough QEMU. 


 I'm sure other examples apply but this at least is an issue. 

the opposite would happen too: if i want to build a aarch64 image for x86_64 target, there would be missing those two because it was intended to be run with g++ native. 

The first solution that comes to mind would be to build a 46x46 matrix of {host}x{target} but that not a good idea.

Anybody has an idea?

**Comment by davidcole1340** [comment]
I started updating the docker images to support AArch64 however have not worked on it for a while: #567

**Comment by allexoll** [comment]
I had a look and it seems that it confirms some of my worries about the complexity of this endeavor. we would need some testing to see if a target-specific compiler would be needed depending on the host arch for each image. I'm also not sure what the space limitation is on dockerhub...

**Comment by Emilgardis** [comment]
> I had a look and it seems that it confirms some of my worries about the complexity of this endeavor. we would need some testing to see if a target-specific compiler would be needed depending on the host arch for each image. I'm also not sure what the space limitation is on dockerhub...

We're moving away from dockerhub, see #609 for switching to GHCR instead

**Comment by nashley** [comment]
Looks like work on this is being tracked here now? https://github.com/cross-rs/cross/issues/751

**Comment by Alexhuszagh** [comment]
Nice catch, this is also partially resolved by #817.
